File: src/main/java/macro303/keschet/Board.java
package macro303.keschet;

import macro303.keschet.pieces.Emperor;
import macro303.keschet.pieces.Piece;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;

public class Board {
	private static final Logger LOGGER = LogManager.getLogger(Board.class);
	private static final int SIZE = 10;

	private Square[][] board = new Square[SIZE][SIZE];

	Board() {
		Arrays.stream(board).forEach(column -> Arrays.setAll(column, row -> new Square()));
	}

	public static int calculateDistance(Pair<Integer, Integer> start, Pair<Integer, Integer> end) {
		Direction direction = calculateDirection(start, end);
		int distance = 0;
		if (direction == Direction.NORTH || direction == Direction.SOUTH)
			distance = start.getY() - end.getY();
		else if (direction == Direction.NORTH_EAST || direction == Direction.EAST || direction == Direction.SOUTH_EAST || direction == Direction.SOUTH_WEST || direction == Direction.WEST || direction == Direction.NORTH_WEST)
			distance = start.getX() - end.getX();
		LOGGER.trace("int calculateDistance(Pair<Integer, Integer>, Pair<Integer, Integer>) = " + distance);
		return distance;
	}

	public static Direction calculateDirection(Pair<Integer, Integer> start, Pair<Integer, Integer> end) {
		Direction direction = Direction.INVALID;
		if (Objects.equals(end.getX(), start.getX()) && end.getY() - start.getY() < 0)
			direction = Direction.NORTH;
		else if (end.getX() - start.getX() > 0 && end.getY() - start.getY() < 0 && Math.abs((end.getX() - start.getX())) - Math.abs((end.getY() - start.getY())) == 0)
			direction = Direction.NORTH_EAST;
		else if (end.getX() - start.getX() > 0 && Objects.equals(end.getY(), start.getY()))
			direction = Direction.EAST;
		else if (end.getX() - start.getX() > 0 && end.getY() - start.getY() > 0 && Math.abs((end.getX() - start.getX())) - Math.abs((end.getY() - start.getY())) == 0)
			direction = Direction.SOUTH_EAST;
		else if (Objects.equals(end.getX(), start.getX()) && end.getY() - start.getY() > 0)
			direction = Direction.SOUTH;
		else if (end.getX() - start.getX() < 0 && end.getY() - start.getY() > 0 && Math.abs((end.getX() - start.getX())) - Math.abs((end.getY() - start.getY())) == 0)
			direction = Direction.SOUTH_WEST;
		else if (end.getX() - start.getX() < 0 && Objects.equals(end.getY(), start.getY()))
			direction = Direction.WEST;
		else if (end.getX() - start.getX() < 0 && end.getY() - start.getY() < 0 && Math.abs((end.getX() - start.getX())) - Math.abs((end.getY() - start.getY())) == 0)
			direction = Direction.NORTH_WEST;
		LOGGER.trace("Direction calculateDirection(Pair<Integer, Integer>, Pair<Integer, Integer>) = " + direction);
		return direction;
	}

	public void draw() {
		for (int column = 0; column <= SIZE; column++) {
			for (int row = 0; row <= SIZE; row++)
				if (column == 0)
					System.out.print((row == 0 ? "" : " ") + row + (row == SIZE ? "" : " "));
				else if (row == 0)
					System.out.print(column + (column == SIZE ? "" : " "));
				else
					Console.showSquare(getSquare(new Pair<>(row, column)).getPiece());
			System.out.println();
		}
	}

	public Square getSquare(Pair<Integer, Integer> square) {
		return square.getX() - 1 < 0 || square.getX() > SIZE || square.getY() - 1 < 0 || square.getY() > SIZE ? null : board[square.getY() - 1][square.getX() - 1];
	}

	public int countPieces(Team team) {
		int counter = 0;
		for (int column = 1; column <= SIZE; column++)
			for (int row = 1; row <= SIZE; row++) {
				Square square = getSquare(new Pair<>(row, column));
				if (square.getPiece() != null && square.getPiece().getTeamColour() == team.getColour() && !(square.getPiece() instanceof Emperor))
					counter++;
			}
		LOGGER.trace("int countPieces(Team) = " + counter);
		return counter;
	}

	public boolean pieceStillOnBoard(Class clazz, Console.Colour teamColour) {
		boolean stillAvailable = false;
		for (int column = 1; column <= SIZE; column++)
			for (int row = 1; row <= SIZE; row++) {
				Square square = getSquare(new Pair<>(row, column));
				if (square.getPiece() != null && square.getPiece().getTeamColour() == teamColour && square.getPiece().getClass() == clazz)
					stillAvailable = true;
			}
		LOGGER.trace("boolean pieceStillOnBoard(Class, Colour) = " + stillAvailable);
		return stillAvailable;
	}

	public ArrayList<Piece> getAllSurroundingPieces(Pair<Integer, Integer> squareLocation) {
		ArrayList<Piece> pieces = new ArrayList<>();
		Square square = getSquare(new Pair<>(squareLocation.getX(), squareLocation.getY() + 1));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX() + 1, squareLocation.getY() + 1));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX() + 1, squareLocation.getY()));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX() + 1, squareLocation.getY() - 1));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX(), squareLocation.getY() - 1));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX() - 1, squareLocation.getY() - 1));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX() - 1, squareLocation.getY()));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		square = getSquare(new Pair<>(squareLocation.getX() - 1, squareLocation.getY() + 1));
		if (square != null && square.getPiece() != null)
			pieces.add(square.getPiece());
		LOGGER.trace("ArrayList<Piece> getAllSurroundingPieces(Pair<Integer, Integer>) = " + pieces);
		return pieces;
	}
}

File: src/main/kotlin/macro303/keschet/Board.kt
package macro303.keschet

import macro303.keschet.pieces.Emperor
import macro303.keschet.pieces.Piece
import org.apache.logging.log4j.LogManager
import java.util.*

class Board {

	private var board = Array(SIZE, { Array(SIZE, { Square() }) })

	fun draw() {
		(0..SIZE).forEach { column ->
			(0..SIZE).forEach {
				when {
					column == 0 -> print("${if (it == 0) "" else " "}$it${if (it == SIZE) "" else " "}")
					it == 0 -> print("$column${if (column == SIZE) "" else " "}")
					else -> Console.showSquare(getSquare(Pair(it, column))!!.piece)
				}
			}
			println()
		}
	}

	fun getSquare(square: Pair<Int, Int>): Square? {
		return if (square.first - 1 < 0 || square.first > SIZE || square.second - 1 < 0 || square.second > SIZE) null else board[square.second - 1][square.first - 1]
	}

	fun countPieces(team: Team): Int {
		var counter = 0
		board.forEach { it -> counter += it.count { it.piece != null && it.piece?.teamColour == team.colour && it.piece !is Emperor } }
		LOGGER.trace("int countPieces(Team) = $counter")
		return counter
	}

	fun pieceStillOnBoard(clazz: Class<*>, teamColour: Console.Colour): Boolean {
		val stillAvailable = board.any { it.any { it.piece != null && it.piece?.teamColour == teamColour && it.piece?.javaClass == clazz } }
		LOGGER.trace("boolean pieceStillOnBoard(Class, Colour) = $stillAvailable")
		return stillAvailable
	}

	fun getAllSurroundingPieces(squareLocation: Pair<Int, Int>): ArrayList<Piece> {
		val pieces = ArrayList<Piece>()
		var square = getSquare(square = Pair(squareLocation.first, squareLocation.second + 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first + 1, squareLocation.second + 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first + 1, squareLocation.second))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first + 1, squareLocation.second - 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first, squareLocation.second - 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first - 1, squareLocation.second - 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first - 1, squareLocation.second))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		square = getSquare(square = Pair(squareLocation.first - 1, squareLocation.second + 1))
		if (square != null && square.piece != null)
			pieces.add(square.piece!!)
		LOGGER.trace("ArrayList<Piece> getAllSurroundingPieces(Pair<Integer, Integer>) = $pieces")
		return pieces
	}

	companion object {
		private val LOGGER = LogManager.getLogger(Board::class.java)
		private val SIZE = 10

		fun calculateDistance(start: Pair<Int, Int>, end: Pair<Int, Int>): Int {
			val distance = when (calculateDirection(start, end)) {
				Direction.NORTH, Direction.SOUTH -> start.second - end.second
				Direction.NORTH_EAST, Direction.EAST, Direction.SOUTH_EAST, Direction.SOUTH_WEST, Direction.WEST, Direction.NORTH_WEST -> start.first - end.first
				else -> 0
			}
			LOGGER.trace("int calculateDistance(Pair<Integer, Integer>, Pair<Integer, Integer>) = $distance")
			return distance
		}

		fun calculateDirection(start: Pair<Int, Int>, end: Pair<Int, Int>): Direction {
			var direction = Direction.INVALID
			when {
				end.first == start.first && end.second - start.second < 0 -> direction = Direction.NORTH
				end.first - start.first > 0 && end.second - start.second < 0 && Math.abs(end.first - start.first) - Math.abs(end.second - start.second) == 0 -> direction = Direction.NORTH_EAST
				end.first - start.first > 0 && end.second == start.second -> direction = Direction.EAST
				end.first - start.first > 0 && end.second - start.second > 0 && Math.abs(end.first - start.first) - Math.abs(end.second - start.second) == 0 -> direction = Direction.SOUTH_EAST
				end.first == start.first && end.second - start.second > 0 -> direction = Direction.SOUTH
				end.first - start.first < 0 && end.second - start.second > 0 && Math.abs(end.first - start.first) - Math.abs(end.second - start.second) == 0 -> direction = Direction.SOUTH_WEST
				end.first - start.first < 0 && end.second == start.second -> direction = Direction.WEST
				end.first - start.first < 0 && end.second - start.second < 0 && Math.abs(end.first - start.first) - Math.abs(end.second - start.second) == 0 -> direction = Direction.NORTH_WEST
			}
			LOGGER.trace("Direction calculateDirection(Pair<Integer, Integer>, Pair<Integer, Integer>) = $direction")
			return direction
		}
	}
}